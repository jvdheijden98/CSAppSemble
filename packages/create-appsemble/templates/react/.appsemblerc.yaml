actions:
  onLoad:
    required: true

output: dist
