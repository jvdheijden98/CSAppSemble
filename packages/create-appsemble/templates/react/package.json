{
  "dependencies": {
    "@appsemble/react": "^1.0.0",
    "@appsemble/sdk": "^1.0.0",
    "prop-types": "^15.6.2",
    "react": "^16.7.0"
  }
}
