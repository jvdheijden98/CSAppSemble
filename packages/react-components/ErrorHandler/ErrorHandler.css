.container {
  margin: 20vh 8px 8px;
  text-align: center;
}
