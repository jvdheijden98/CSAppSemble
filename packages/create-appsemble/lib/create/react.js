import fs from 'fs-extra';
import path from 'path';
import sortPackageJson from 'sort-package-json';

import { dependencies } from '../../templates/react/package.json';

export default async function react(outputPath, pkg) {
  const outPkg = sortPackageJson({
    ...pkg,
    dependencies: { ...pkg.dependencies, ...dependencies },
  });

  await fs.copy(path.resolve(__dirname, '../../templates/react'), outputPath);
  await fs.outputJSON(path.join(outputPath, 'package.json'), outPkg, { spaces: 2 });
}
