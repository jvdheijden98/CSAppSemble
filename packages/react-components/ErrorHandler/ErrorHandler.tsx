import { captureException, withScope } from '@sentry/browser';
import * as React from 'react';

export interface ErrorHandlerProps {
  children: React.ReactNode;
  fallback: React.ElementType;
}

interface ErrorHandlerState {
  error: boolean;
}

/**
 * Capture renderer errors using Sentry.
 */
export default class ErrorHandler extends React.Component<ErrorHandlerProps, ErrorHandlerState> {
  state: ErrorHandlerState = {
    error: false,
  };

  componentDidCatch(error: Error, info: React.ErrorInfo): void {
    this.setState({ error: true });
    withScope(scope => {
      Object.entries(info).forEach(([key, value]) => {
        scope.setExtra(key, value);
      });
      captureException(error);
    });
  }

  render(): React.ReactNode {
    const { children, fallback: Fallback } = this.props;
    const { error } = this.state;

    return error ? <Fallback /> : children;
  }
}
