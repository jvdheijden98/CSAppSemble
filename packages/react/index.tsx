import { bootstrap as sdkBootstrap, BootstrapParams } from '@appsemble/sdk';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import retargetEvents from 'react-shadow-dom-retarget-events';

export interface BlockProps<P = any, A = {}> extends BootstrapParams<P, A> {
  /**
   * The DOM node on which the block is mounted.
   */
  reactRoot: HTMLElement;
}

const { Consumer, Provider } = React.createContext<BlockProps>(null);

/**
 * Mount a React component returned by a bootstrap function in the shadow DOM of a block.
 */
export function mount<P, A = {}>(
  Component: React.ComponentType<BlockProps<P, A>>,
  root?: HTMLElement,
): (params: BootstrapParams<P, A>) => void {
  return params => {
    const reactRoot = params.shadowRoot.appendChild(
      root ? root.cloneNode() : document.createElement('div'),
    ) as HTMLElement;
    const props = {
      ...params,
      reactRoot,
    };
    const component: JSX.Element = (
      <Provider value={props}>
        <Component {...props} />
      </Provider>
    );
    ReactDOM.render(component, reactRoot);
    params.utils.addCleanup(() => ReactDOM.unmountComponentAtNode(reactRoot));
    /**
     * React doesnâ€™t play nice with shadow DOM. This library works around that. However, the
     * implementation does contain some bugs.
     *
     * Related issues:
     *
     * - https://github.com/facebook/react/issues/2043
     * - https://github.com/facebook/react/issues/9242
     * - https://github.com/facebook/react/issues/10422
     * - https://github.com/facebook/react/issues/12973
     * - https://github.com/facebook/react/pull/8117
     * - https://github.com/facebook/react/pull/12163
     *
     * @todo Remove this when React has fixes this issue.
     */
    retargetEvents(params.shadowRoot);
  };
}

export function bootstrap<P, A = {}>(
  Component: React.ComponentType<BlockProps<P, A>>,
  reactRoot?: HTMLElement,
): void {
  sdkBootstrap<P, A>(mount(Component, reactRoot));
}

/**
 * A HOC which passes the Appsemble block values to he wrapped React component.
 */
export function withBlock<P extends object>(
  Component: React.ComponentType<P & Omit<BlockProps, keyof P>>,
): React.ComponentType<P> {
  function Wrapper(props: P): JSX.Element {
    return <Consumer>{values => <Component {...values} {...props} />}</Consumer>;
  }
  if (process.env.NODE_ENV !== 'production') {
    Wrapper.displayName = `withBlock(${Component.displayName || Component.name})`;
  }
  return Wrapper;
}
