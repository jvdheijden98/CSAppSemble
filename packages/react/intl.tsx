import * as React from 'react';
import { IntlConfig, IntlProvider } from 'react-intl';

import { BlockProps } from '.';

/**
 * A HOC which provides `intl` in a block written in React.
 */
// eslint-disable-next-line import/prefer-default-export
export function provideIntl<P = any, A = {}>(
  Component: React.ComponentType<BlockProps<P, A>>,
  intlProviderProps?: Partial<IntlConfig>,
): React.ComponentType<BlockProps<P, A>> {
  return props => (
    <IntlProvider defaultLocale="en-US" locale="en-US" {...intlProviderProps}>
      <Component {...props} />
    </IntlProvider>
  );
}
