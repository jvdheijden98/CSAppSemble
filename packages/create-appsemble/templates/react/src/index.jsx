import { bootstrap } from '@appsemble/react';
import { provideIntl } from '@appsemble/react/intl';

import TableBlock from './components/TableBlock';

bootstrap(provideIntl(TableBlock));
