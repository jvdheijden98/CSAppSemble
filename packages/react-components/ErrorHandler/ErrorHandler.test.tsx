import { mount } from 'enzyme';
import * as React from 'react';

import ErrorHandler from './ErrorHandler';

it('should render its children if no errors are thrown', () => {
  const Child = (): JSX.Element => <p>Test</p>;
  const Fallback = (): JSX.Element => <p>Something went wrong!</p>;
  const result = mount(
    <ErrorHandler fallback={Fallback}>
      <Child />
    </ErrorHandler>,
  );

  expect(result).toMatchSnapshot();
});

it('should render its fallback when errors are thrown', () => {
  jest.spyOn(console, 'error').mockImplementation();
  const Child = (): JSX.Element => {
    throw Error('test');
  };
  const Fallback = (): JSX.Element => <p>Something went wrong!</p>;
  const result = mount(
    <ErrorHandler fallback={Fallback}>
      <Child />
    </ErrorHandler>,
  );

  expect(result).toMatchSnapshot();
});
