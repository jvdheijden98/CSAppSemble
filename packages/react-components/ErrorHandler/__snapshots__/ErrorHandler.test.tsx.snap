// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render its children if no errors are thrown 1`] = `
<ErrorHandler
  fallback={[Function]}
>
  <Child>
    <p>
      Test
    </p>
  </Child>
</ErrorHandler>
`;

exports[`should render its fallback when errors are thrown 1`] = `
<ErrorHandler
  fallback={[Function]}
>
  <Fallback>
    <p>
      Something went wrong!
    </p>
  </Fallback>
</ErrorHandler>
`;
